### PR TITLE
Add gate definitions and language selection for Native Gates

### PR DESCRIPTION
**Pull request by Cynocracy**
IonQ Native gates support in Qiskit

Bump to 0.3.0

### Summary

This PR allows configuration of a 'lang' when creating an IonQBackend, which controls the basis gates used. "qis" is the existing default, and "native" allows for programming with GPI, GPI2 and MS gates.

These gates are also defined in this PR, along with matrix representations

### CODE REVIEW

**Comment by Cynocracy** [comment]
Released this as v0.3.0, freaked a bit, yanked it, and then again as v0.3.0.dev, haha

So; no rush on review until we want to do a proper semver update (which will have to be 0.3.1 now, whoops!)

File: qiskit_ionq/ionq_gates.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+class MSGate(Gate):
+    r"""Entangling 2-Qubit MS gate.
+    **Circuit symbol:**
+    .. parsed-literal::
+              _______
+        q_0: ┤       ├-
+             |MS(ϴ,0)|
+        q_1: ┤       ├-
+             └───────┘
```
Comment (ColemanCollins) [inline qiskit_ionq/ionq_gates.py:99]: phi one phi two, no?

File: qiskit_ionq/helpers.py [hunk 2]
```diff
@@ -325,10 +369,11 @@ def qiskit_to_ionq(circuit, backend_name, passed_args=None):
         "target": backend_name[5:],
         "shots": passed_args.get("shots"),
         "body": {
+            "gateset": gateset,
             "qubits": circuit.num_qubits,
             "circuit": ionq_circ,
         },
-        "registers": {"meas_mapped": meas_map},
+        "registers": {"meas_mapped": meas_map} if meas_map else {},
```
Comment (ColemanCollins) [inline qiskit_ionq/helpers.py:376]: :nod:

File: qiskit_ionq/ionq_job.py [hunk 3]
```diff
@@ -93,10 +93,14 @@ def _build_counts(result, use_sampler=False, sampler_seed=None):
...
     shots = metadata.get("shots", "1024")
-    shots = int(shots) if shots.isdigit() else 1024  # We do this in case shots was 0 or None.
+    shots = (
+        int(shots) if shots.isdigit() else 1024
+    )  # We do this in case shots was 0 or None.
 
     # Grab the mapped output from response.
-    output_probs = result["data"].get("registers", {}).get("meas_mapped", {})
+    output_probs = (result["data"].get("registers", {}) or {}).get("meas_mapped", {})
+    if not output_probs:
```
Comment (ColemanCollins) [inline qiskit_ionq/ionq_job.py:102]: :nod intensifies:

File: test/helpers/test_qiskit_to_ionq.py [hunk 4]
```diff
@@ -190,3 +199,134 @@ def test_full_circuit(simulator_backend):
     assert actual_metadata_header == expected_metadata_header
     assert actual_output_map == expected_output_map
     assert actual == expected_rest_of_payload
+
+
+def test_circuit_transpile(simulator_backend):
```
Comment (ColemanCollins) [inline test/helpers/test_qiskit_to_ionq.py:204]: love to assert the error cases. brings joy to my cold heart.

**Review by ColemanCollins — APPROVED** [review]
I don't necessarily trust my ability to spot-check these gate unitaries and their numpy array representations — might be nice to get someone with "real" QIS experience to have a look at that — but otherwise this looks great. Thanks again for jumpin' on it!

File: test/helpers/test_qiskit_to_ionq.py (hunk 4 above)
Comment (Cynocracy) [reply test/helpers/test_qiskit_to_ionq.py:204]: :100:

File: qiskit_ionq/ionq_gates.py (hunk 1 above)
Comment (Cynocracy) [reply qiskit_ionq/ionq_gates.py:99]: it's 0 indexed now :)

**Pull request merged**